### PR TITLE
dashboard: So Long

### LINKED ISSUE #8994: [Feature request] Dashboard - add "Show all" tab to view all graphs
Would be nice if dashboard had option to show all graphs at once. Now i have to "display:block;" each tab manually in browser dev tools

Preview:
dashboard_show_all

### PR DESCRIPTION
**Pull request by marusak**
This drops dashboard. I plan to discuss this on the Wednesday call, currently I have 3 ideas how we can proceed:
- Just drop it. We need to also make sure that it is fine with released packages. (this PR stands)
- Move it to other repo. We need to get new dist-gits. (this PR stands)
- Keep the package but implement there just graphs. Rename it to something else or possibly don't show it as normal menu item but rather as 'Graphs' in the host switcher. (This PR needs additional work and would most likely depend on https://github.com/cockpit-project/cockpit/pull/14850 and would fix #8994)

Other ideas welcomed :)

### CODE REVIEW

File: tools/debian/control
Comment (martinpitt) [inline tools/debian/control:142]: Needs a `Breaks: cockpit-dashboard` as well; a single Replaces: only means "replaces some files", but with the extra Breaks, apt will properly remove the old package on upgrades.

**Comment by andreasn** [comment]
As discussed on Wednesday, let's just drop it. For anyone who wants multi-host graphs, there is Grafana that does a lot better job at visualizing perfomance.
We need to make Grafana easier to install though.

**Comment by marusak** [comment]
Setting to blocked until we resolve RCM ticket in rhel.

**Comment by marusak** [comment]
I got green to proceed with this from RCM.
@mvollmer WDYT since you are looking into graphs still?

**Comment by marusak** [comment]
Ah, also needs some attention in bots repository.

**Comment by marusak** [comment]
see https://github.com/cockpit-project/bots/pull/1423

**Comment by mvollmer** [comment]
> I got green to proceed with this from RCM.
> @mvollmer WDYT since you are looking into graphs still?

Please go ahead with removing the dashboard. Personally, I am happy to see it go, I was just afraid it would be rude to our users.

File: tools/debian/control
Comment (martinpitt) [inline tools/debian/control:68]: These replaces/breaks must stay for a few years until 233 is in the oldest supported Debian/Ubuntu stable.

File: tools/cockpit.spec
Comment (martinpitt) [inline tools/cockpit.spec:368]: This is unrelated, but ok -- we don't support any RHEL or other distro any more with such an old cockpit.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Looks great, thanks! Just a minor packaging issue.

File: tools/debian/control
Comment (marusak) [reply tools/debian/control]: Good point. `cockpit-dashboard` was replaced twice, so just kept one.

**Review by martinpitt — APPROVED** [review]
Cheers!

**Comment by martinpitt** [comment]
I suggest to massaging the commit message a little when landing this.

**Comment by marusak** [comment]
Damn, conflicts

**Review by martinpitt — APPROVED** [review]

**Comment by didib** [comment]
Hi, would you consider adding 'Provides: cockpit-dashboard'?

I have an oVirt engine machine on CentOS 8.3, and ugprading to Stream fails, because ovirt-cockpit-sso Requires: cockpit-dashboard. I pushed [1] to fix this, but perhaps it would help to also add Provides: . I don't mind opening a new issue for this if you want.

[1] https://github.com/oVirt/ovirt-cockpit-sso/pull/30

**Comment by martinpitt** [comment]
It would be factually wrong, as c-system really does not provide the old "Dashboard". Maybe someone wants to maintain it as a separate project or install from an older repo, and then these would collide?

FTR: Curiously the `Obsoletes:` seems to even be too weak: https://bugzilla.redhat.com/show_bug.cgi?id=1911825 So I wonder why the upgrade fails, I would have expected dnf to just keep the cockpit-dashboard around then?

**Comment by didib** [comment]
> It would be factually wrong, as c-system really does not provide the old "Dashboard". Maybe someone wants to maintain it as a separate project or install from an older repo, and then these would collide?

I see your point.

Perhaps clarify why you added Obsoletes:, then.

Is there a risk/problem with having it installed? If not, why not just ignore? If it's "just" to get rid of garbage, perhaps you better recommend users do 'dnf autoremove SOMETHING'.

> 
> FTR: Curiously the `Obsoletes:` seems to even be too weak: https://bugzilla.redhat.com/show_bug.cgi?id=1911825 So I wonder why the upgrade fails, I would have expected dnf to just keep the cockpit-dashboard around then?

I think Obsoletes: prevents that, no?

**Comment by martinpitt** [comment]
> Perhaps clarify why you added Obsoletes:, then.

My main concern was to get the obsolete package cleaned up automatically on upgrade. That's best practice at least in the Debian world (where I learned packaging), to avoid building up cruft on upgrades. Also, as long as there is no official replacement, the old package will just bitrot and eventually fail to work, as it's not being tested any more.

As to *how* exactly declare that in the .spec, I'm all ears for a better way. For Debian it is `Breaks:/Replaces: <oldpackage>` (which we used), but I find it rather difficult to find documentation as  to how to do that in RPM land. If `Provides:` helps with that, then let's add it (it just sounds a bit weird, and like a false promise -- but maybe that's just bad intuition with dnf/rpm?)

Thanks!

**Comment by didib** [comment]
> > Perhaps clarify why you added Obsoletes:, then.
> 
> My main concern was to get the obsolete package cleaned up automatically on upgrade. That's best practice at least in the Debian world (where I learned packaging), to avoid building up cruft on upgrades. Also, as long as there is no official replacement, the old package will just bitrot and eventually fail to work, as it's not being tested any more.

I am by no means am rpm/dnf expert, myself, nor am I aware of a comprehensive analysis of all the relevant flows and how rpm/dnf should behave in each, what users can/should do, etc. That said, I think it goes like this:

There is an option clean_requirements_on_remove, True by default, and if it's set, 'dnf remove package' will remove also dependencies installed by it that are not needed by other installed packages (read the docs for more details). Not sure how this applies to 'dnf update' - can't quickly find a definitive answer. From a quick look at the code, I think it does not. Perhaps that's worth a bug.

In addition, there is a command "dnf autoremove" that removes all such dependencies.

> 
> As to _how_ exactly declare that in the .spec, I'm all ears for a better way. For Debian it is `Breaks:/Replaces: <oldpackage>` (which we used), but I find it rather difficult to find documentation as to how to do that in RPM land. If `Provides:` helps with that, then let's add it (it just sounds a bit weird, and like a false promise -- but maybe that's just bad intuition with dnf/rpm?)

I *think* that "Provides:" is somewhat similar to Debian's "Replaces:". Also "Replaces" might sound like the broken promise you mention, if you are not used to how it works.

Perhaps one way to think about this is: If some package p0, at some point, added to its spec 'Requires: p1', and now I am going to stop supporting p1, and suggest/recommend that users remove it, even to the extent that I declare some other package p2 to Obsolete: p1, then what do I expect the maintainers of p0 to do? What you now chose is:

1. I want things to break, so that users will notice, and require the maintainers to decide what to do. If they do not really need p1, just remove the Requires:. If they actually want p2, Requires: p2. If they really require p1, they should adopt and continue maintaining it, and ask me to remove the Obsoletes: in p2, if there is a reasonable use case to having both installed.

Another option is:

2. I want things to be as smooth as possible - to make it so that installing/updating p2 will just remove p1, and that if p0 does not really need p1, nothing will be broken. If it does, things will be broken anyway, but there is nothing that can prevent that.

I think choosing between them is both a matter of policy, but also a matter of personal opinion on how things should work :-).

In practice, I never was in your position. The only cases where I added/removed Obsoletes:/Provides: were for packages that were (only?) relevant to my project (oVirt) and that I knew quite well what other packages might Require: them. So if you consider cockpit a kind of library, in this regard, then sadly I can't tell what's the best option. In my original comment I suggested basically (2.), mainly because I wasn't aware of the details and thought there is no real need for requiring every other project to explicitly change (like we now did in ovirt-cockpit-sso). Generally, I agree that (1.) might be better, at least if you think there is a good chance that you'll rather quickly find (or force) all the users to notice and fix.

So: If you think there is a good chance that:

1. Some package now Requires: cockpit-dashboard
2. If you now quietly remove it by adding "Provides:", it will make things break,
3. But not immediately, nor in every case - rather, only occasionally, in rather hard-to-debug ways

Then I think you made the right choice. Better cause breakage immediately by dnf then intermittently later in actual application code. Otherwise, perhaps do add Provides:.

Sorry for such a long reply. HTH. Thanks!

**Pull request merged**
